### PR TITLE
Block operations on chunks

### PR DESCRIPTION
**Pull request by cevian**
This PR adds logic to block alter table and drop operations on chunks.

### CODE REVIEW

File: src/process_utility.c [hunk 1]
```diff
@@ -47,6 +53,34 @@ prev_ProcessUtility(Node *parsetree,
 	}
 }
 
+static void
+verify_not_chunk(Oid relid, bool allow_overide)
```
Comment (olofr) [inline src/process_utility.c]: I think this function name is confusing. Something more like `check_chunk_operation_allowed` maybe ?

File: sql/chunk.sql [hunk 2]
```diff
@@ -28,6 +34,8 @@ BEGIN
     -- is executed as a result of a DROP TABLE on the hypertable
     -- that this chunk belongs to.
 
+    PERFORM _timescaledb_internal.chunk_allow_modification(TRUE);
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/chunk.sql]: After further review, it also seems weird that this GUC variable is even needed here. The only thing this function should do is to remove the metadata from the `chunk_constraint` table. The need for the GUC variable seems to come from the fact that you are doing an `ALTER TABLE` in `drop_chunk_constraint()` in order to remove the constraints from the chunk -- something which shouldn't be needed if you anyhow are dropping the table. This indicates that there isn't a proper separation of concerns here. There should be one function that drops the metadata and one that does the alter table stuff. Only the former should be called when dropping the chunk table. If needed, you can have a wrapper that does both as well.

[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/chunk.sql]: Also, I see no reason we should block manually dropping a chunk. Is there a good reason for doing so? Or, is this simply a side-effect of you doing also `ALTER TABLE` when dropping a chunk (as described in the comment above)?

[the remark this answers is not in the document]
Comment (cevian) [reply sql/chunk.sql]: Note this has nothing to do with the GUC variable. This is setting expect_chunk_modification variable in process_utility, which gets reset on transaction aborts.

File: src/process_utility.c (hunk 1 above)
Comment (cevian) [reply src/process_utility.c]: fixed

File: src/chunk.c [hunk 3]
```diff
@@ -663,3 +665,87 @@ chunk_copy(Chunk *chunk)
...
+	 */
+	ScanKeyInit(&scankey[0], Anum_chunk_schema_name_idx_schema_name, BTEqualStrategyNumber,
+				F_NAMEEQ, CStringGetDatum(schema_name));
+
+	ScanKeyInit(&scankey[1], Anum_chunk_schema_name_idx_table_name, BTEqualStrategyNumber,
+				F_NAMEEQ, CStringGetDatum(table_name));
+
+	num_found = scanner_scan(&ctx);
+
+	if (num_found > 1)
```
[the remark this answers is not in the document]
Comment (cevian) [reply src/chunk.c]: fixed

File: src/chunk.c [hunk 4]
```diff
@@ -663,3 +665,87 @@ chunk_copy(Chunk *chunk)
...
+		return chunk;
+	}
+
+	return NULL;
+}
+
+bool
+chunk_exists(const char *schema_name, const char *table_name)
+{
+	Chunk	   *chunk = chunk_get_by_name(schema_name, table_name, 0, false);
```
[the remark this answers is not in the document]
Comment (cevian) [reply src/chunk.c]: fixed

File: src/guc.c [hunk 5]
```diff
@@ -51,6 +52,16 @@ _guc_init(void)
 							 NULL,
 							 NULL,
 							 NULL);
+	DefineCustomBoolVariable("timescaledb.allow_alter_chunk", "Allow altering chunk tables directly",
```
[the remark this answers is not in the document]
Comment (cevian) [reply src/guc.c]: removed GUC

File: src/process_utility.c [hunk 6]
```diff
@@ -47,6 +53,34 @@ prev_ProcessUtility(Node *parsetree,
...
+		return;
+
+	if (chunk_exists_relid(relid))
+	{
+		if (allow_overide)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("Operation discouraged on chunk tables."),
+			  errhint("We do not recommend altering chunk tables directly.\n"
```
[the remark this answers is not in the document]
Comment (cevian) [reply src/process_utility.c]: removed GUC

File: sql/chunk.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (cevian) [reply sql/chunk.sql]: removed GUC; handling (not blocking) drop chunks.

**Comment by cevian** [comment]
@olofr @erimatnor Please review this PR again. I removed the GUC and am now handling and not blocking chunk drops.

File: src/chunk.c [hunk 7]
```diff
@@ -663,3 +665,80 @@ chunk_copy(Chunk *chunk)
...
+
+	ScanKeyInit(&scankey[1], Anum_chunk_schema_name_idx_table_name, BTEqualStrategyNumber,
+				F_NAMEEQ, CStringGetDatum(table_name));
+
+	num_found = scanner_scan(&ctx);
+
+	switch(num_found) 
+	{
+		case 0:
+			if(fail_if_not_found)
```
Comment (erimatnor) [inline src/chunk.c]: missing whitespace

File: src/chunk.c [hunk 8]
```diff
@@ -663,3 +665,80 @@ chunk_copy(Chunk *chunk)
...
+	 */
+	ScanKeyInit(&scankey[0], Anum_chunk_schema_name_idx_schema_name, BTEqualStrategyNumber,
+				F_NAMEEQ, CStringGetDatum(schema_name));
+
+	ScanKeyInit(&scankey[1], Anum_chunk_schema_name_idx_table_name, BTEqualStrategyNumber,
+				F_NAMEEQ, CStringGetDatum(table_name));
+
+	num_found = scanner_scan(&ctx);
+
+	switch(num_found)
```
Comment (erimatnor) [inline src/chunk.c]: missing whitespace

**Pull request merged**